### ISSUE TITLE
"Unable to replace folder" error when downloading some maps

### ISSUE THREAD

**Issue by JanSimek** [triplea-game/triplea#11167]
### Reproduction Steps
1. unzip triplea-game-headed.zip and run `java -jar bin/triplea-game-headed-2.6+14192.jar`
2. click Download maps and select world_war_ii_v5_1942 or middle_earth
3. see a popup: "Unable to replace folder: /home/jsimek/triplea/downloadedMaps/world_war_ii_v5_1942 <br/>Are you low on disk space?<br/> Is the destination folder owned by administrator but you are running TripleA as a non-administrator?"

The above error does not happen with the "1941" map, for example.

### Engine Version

Engine Version: 2.6+14192
OS: Arch Linux
Java: openjdk version "19.0.1" 2022-10-18

### Is there anything else we should know?

```
11:10:34.970 [main] INFO  o.t.game.client.HeadedGameRunner - Launching game, version: 2.6+14192 
11:10:56.421 [Thread-10] INFO  g.s.e.f.m.f.s.l.ZippedMapsExtractor - Extracting map zip: /tmp/triplea11291612084988160259/1941.zip -> /home/jsimek/triplea/downloadedMaps/1941
11:11:18.634 [Thread-14] INFO  g.s.e.f.m.f.s.l.ZippedMapsExtractor - Extracting map zip: /tmp/triplea3726742665895803946/world_war_ii_v5_1942.zip -> /home/jsimek/triplea/downloadedMaps/world_war_ii_v5_1942
11:11:18.669 [Thread-15] INFO  g.s.e.f.m.f.s.l.ZippedMapsExtractor - Extracting map zip: /tmp/triplea2595514499941526996/world_war_ii_v5_1942.zip -> /home/jsimek/triplea/downloadedMaps/world_war_ii_v5_1942
11:11:18.771 [Thread-15] WARN  org.triplea.io.FileUtils - Unable to replace folder: /home/jsimek/triplea/downloadedMaps/world_war_ii_v5_1942 <br/>Are you low on disk space?<br/> Is the destination folder owned by administrator but you are running TripleA as a non-administrator?
11:11:18.834 [Thread-15] ERROR org.triplea.io.FileUtils - Failed to rollback, failed to restore backup folder: /tmp/temp-dir12671110851960406886/world_war_ii_v5_1942, to: /home/jsimek/triplea/downloadedMaps/world_war_ii_v5_1942
```

```
❯ tree -pufid /home/jsimek/triplea
[drwxrwxrwx jsimek  ]  /home/jsimek/triplea
[drwxrwxrwx jsimek  ]  /home/jsimek/triplea/downloadedMaps
[drwx------ jsimek  ]  /home/jsimek/triplea/downloadedMaps/triplea-unzip13771364541106472434
[drwxr-xr-x jsimek  ]  /home/jsimek/triplea/downloadedMaps/1941
[drwxr-xr-x jsimek  ]  /home/jsimek/triplea/downloadedMaps/1941/map
[drwxr-xr-x jsimek  ]  /home/jsimek/triplea/downloadedMaps/1941/map/baseTiles
[drwxr-xr-x jsimek  ]  /home/jsimek/triplea/downloadedMaps/1941/map/flags
[drwxr-xr-x jsimek  ]  /home/jsimek/triplea/downloadedMaps/1941/map/games
[drwxr-xr-x jsimek  ]  /home/jsimek/triplea/downloadedMaps/1941/map/misc
[drwxr-xr-x jsimek  ]  /home/jsimek/triplea/downloadedMaps/1941/map/reliefTiles
[drwxr-xr-x jsimek  ]  /home/jsimek/triplea/downloadedMaps/1941/map/units
[drwxr-xr-x jsimek  ]  /home/jsimek/triplea/downloadedMaps/1941/map/units/Americans
[drwxr-xr-x jsimek  ]  /home/jsimek/triplea/downloadedMaps/1941/map/units/British
[drwxr-xr-x jsimek  ]  /home/jsimek/triplea/downloadedMaps/1941/map/units/FreeFrench
[drwxr-xr-x jsimek  ]  /home/jsimek/triplea/downloadedMaps/1941/map/units/Germans
[drwxr-xr-x jsimek  ]  /home/jsimek/triplea/downloadedMaps/1941/map/units/Chinese
[drwxr-xr-x jsimek  ]  /home/jsimek/triplea/downloadedMaps/1941/map/units/Italians
[drwxr-xr-x jsimek  ]  /home/jsimek/triplea/downloadedMaps/1941/map/units/Japanese
[drwxr-xr-x jsimek  ]  /home/jsimek/triplea/downloadedMaps/1941/map/units/Neutral
[drwxr-xr-x jsimek  ]  /home/jsimek/triplea/downloadedMaps/1941/map/units/Russians
[drwxr-xr-x jsimek  ]  /home/jsimek/triplea/downloadedMaps/1941/map/units/VichyFrench
```

**Comment by beelee1** [comment]
@JanSimek are there other maps where it does or doesn't throw the error ?

**Comment by JanSimek** [comment]
@beelee1 it seems most if not all the other maps work. Only the ones I actually want to play cause errors. So far I tested these:

| **Works**                         | **Error**            |
|-----------------------------------|----------------------|
| 1941                              | middle_earth         |
| 1914_global_empires               | world_war_ii_v5_1942 |
| 1939_pack_of_steel_big_world      |                      |
| 2020                              |                      |
| 270bc_tertiered                   |                      |
| china_vs_russia                   |                      |
| zombieland                        |                      |
| world_war_1_end_of_empires        |                      |
| world_war_i_1914                  |                      |
| world_war_ii_classic              |                      |
| world_war_2010                    |                      |
| world_war_ii_v6_1941              |                      |
| world_war_ii_revised_variations   |                      |
| world_war_ii_v3_1941_balanced_mod |                      |
| world_war_ii_v4                   |                      |
| world_war_ii_global               |                      |
| total_world_war                   |                      |
| modern_warfare                    |                      |
| game_of_thrones_v2                |                      |

The ones that cause errors are displayed as being downloaded twice:
![image](https://user-images.githubusercontent.com/1261493/198223253-2e8e535c-19f2-4b83-9588-053f2a05b0f1.png)

**Comment by RoiEXLab** [comment]
I had a look at the code and I'm pretty sure that the fact that they are being downloaded twice is the issue. (Not sure what causes it yet though.)
The code downloads the zip, tries to extract it to the target folder, tries to extract it again into the same folder, but notices this folder is already full, so it tries to remove the existing files first and something in this process seems to go wrong.

@DanVanAtta As a side-note, you implemented a unit test for this and for some reason I can't explain it doesn't seem to cover this case even though it seems to do so, not sure what's going on there maybe you have an idea.

Anyways @JanSimek I'm pretty sure the issue is on our end.
When looking at the map listing I found this:


```json
  {
    "downloadUrl": "https://github.com/triplea-maps/middle_earth/archive/refs/heads/master.zip",
    "previewImageUrl": "https://github.com/triplea-maps/middle_earth/blob/master/preview.png?raw=true",
    "mapName": "middle_earth",
    "lastCommitDateEpochMilli": 1648557975000,
    "description": "<br>Lord of the Rings\n",
    "mapTags": [],
    "downloadSizeInBytes": 15069025
  },
  {
    "downloadUrl": "https://github.com/triplea-maps/lord_of_the_rings_war_of_the_ring/archive/refs/heads/master.zip",
    "previewImageUrl": "https://github.com/triplea-maps/lord_of_the_rings_war_of_the_ring/blob/master/preview.png?raw=true",
    "mapName": "middle_earth",
    "lastCommitDateEpochMilli": 1664920978000,
    "description": "<br>Lord of the Rings\n",
    "mapTags": [],
    "downloadSizeInBytes": 15976350
  },
```

as well as this:
```json
  {
    "downloadUrl": "https://github.com/triplea-maps/world_war_ii_v5_1942/archive/refs/heads/master.zip",
    "previewImageUrl": "https://github.com/triplea-maps/world_war_ii_v5_1942/blob/master/preview.png?raw=true",
    "mapName": "world_war_ii_v5_1942",
    "lastCommitDateEpochMilli": 1631771018000,
    "description": "<br>Second edition update of the popular \"Spring 1942\" game (ww2v4), which itself was an update of \"Revised\" (ww2v2).\n<br>\n<br>Changes:\n<br>1. Several territories/SZs added or modified.\n<br>2. More initial starting units (including new Factories on Karelia and India).\n<br>3. Armor cost increased to 6 PUs.\n<br>4. AA Guns now cost 5 PUs and each can fire at a maximum of 3 planes, each plane being only fired upon once, and they can be combat casualties.\n<br>5. Factories now have their own AA to defend during strategic bombing.\n<br>6. Honolulu is now a Victory City.\n<br>\n",
    "mapTags": [],
    "downloadSizeInBytes": 6789004
  },
  {
    "downloadUrl": "https://github.com/triplea-maps/ww2v5_variants/archive/refs/heads/master.zip",
    "previewImageUrl": "https://github.com/triplea-maps/ww2v5_variants/blob/master/preview.png?raw=true",
    "mapName": "world_war_ii_v5_1942",
    "lastCommitDateEpochMilli": 1657399466000,
    "description": "<br>Second edition update of the popular \"Spring 1942\" game (ww2v4), which itself was an update of \"Revised\" (ww2v2).\n<br>\n<br>Changes:\n<br>1. Several territories/SZs added or modified.\n<br>2. More initial starting units (including new Factories on Karelia and India).\n<br>3. Armor cost increased to 6 PUs.\n<br>4. AA Guns now cost 5 PUs and each can fire at a maximum of 3 planes, each plane being only fired upon once, and they can be combat casualties.\n<br>5. Factories now have their own AA to defend during strategic bombing.\n<br>6. Honolulu is now a Victory City.\n<br>\n",
    "mapTags": [],
    "downloadSizeInBytes": 6790084
  },
```

As you can see there are multiple entries with the same value for `mapName`. The client side code assumes that this name is unique among all maps, that's why it downloads the thing twice (technically it downloads 2 maps which happen to have the same name).

TL;DR There seem to be 2 issues that in combination are causing the problem here:
1. mapName is not unique among all map names, even though the client assumes it is.
2. For some reason the client errors out when downloading 2 maps with the same name instead of replacing each other without notice


So @JanSimek what can you do about this? I'm pretty sure the fastest solution for you would be to download the maps by hand. You can find the direct download link in the JSON I just posted, not sure which one you want to play and how they differ. Then just place the zip inside the `/home/jsimek/triplea/downloadedMaps/` folder and make sure any folders with the same name are gone. This might already do the trick once you restart TripleA. Please report back if this doesn't work, there might be a step I'm missing.

**Comment by RoiEXLab** [comment]
@DanVanAtta What's the best way to fix the duplicate name issue in your opinion? Would a database constraint help to fix such a "silent problem" in the future? Or should be adjust the client to be able to deal with overlapping names?

**Comment by DanVanAtta** [comment]
It might be best for us to use the repo slug name for the extraction folder rather than the 'map name' (maybe we do not need 'map name' at all?)

A UK constraint would not be bad but we would need to be sure that we can continue indexing maps. Though, then we'd have maps that "mysteriously" do not get indexed. 

Requiring a unique map name does seem like a liability. I'm not sure whether we should work with it, or try to constrain it.

**Comment by asvitkine** [comment]
Looks like we did switch to "repo slug name", so I assume this is fixed now.
